File: math_problems/hydra_game/java/com/themaskedcrusader/maths/the_hydra_game/model/Hydra.java
package com.themaskedcrusader.maths.the_hydra_game.model;

public class Hydra {
    Node head;
    Node tail;

    public Hydra(int length) {
        this.head = new Node(1);
        Node node1 = head;
        Node node2 = null;
        for (int i = 1; i < length; i++) {
            node2 = new Node(1);
            node1.next = node2;
            node2.previous = node1;
            node1 = node2;
        }

        tail = node2;
    }

    public boolean areAllHeadsAlone() {
        Node node = head;
        while (node != null) {
            if (node.heads != 1) return false;
            node = node.next;
        }
        return true;
    }

    public void removeHead(Node headNode) {
        Counter.steps++;
        Node parent = headNode.next;
        headNode.heads--;
        if (parent != null) {
            parent.heads += Counter.steps;
        }
    }

    public Node getHead() {
        return head;
    }

    public Node getHeadWithMostHeads() {
        Node checkNode = head;
        Node withMostHeads = null;
        int mostHeadsCount = 0;
        while (checkNode != null) {
            if (checkNode.heads > mostHeadsCount) {
                mostHeadsCount = checkNode.heads;
                withMostHeads = checkNode;
            }
            checkNode = checkNode.next;
        }
        return withMostHeads;
    }

    public void lopHead() {
        if (head == tail) {
            head = null;
            tail = null;
            return;
        }

        Counter.steps++;

        if (head.heads != 1) {
            System.err.println("Can't lop a head with remaining heads");
            System.exit(1);
        }
        Node newHead = head.next;
        newHead.previous = null;
        head = newHead;
        Node next = head.next;
        if (next != null) {
            next.heads += Counter.steps;
        }
    }

    public boolean hasHeads() {
        return head != null && tail != null;
    }
}

File: math_problems/hydra_game/cpp/hydra.cpp
#include <iostream>

struct node {
    unsigned long long int heads = 1;
    node *next;
};

void insert_at_front(node** head, int heads) {
    node* new_node = new node();
    new_node->heads = heads;

    new_node->next = (*head);

    (*head) = new_node;
}

bool are_all_heads_one(node** head) {
    auto check_node = (*head);
    while (check_node != nullptr) {
        if (check_node->heads != 1) return false;
        check_node = check_node->next;
    }
    return true;
}

void lop_head(node** head, unsigned long long int counter) {
    if ((*head)->heads != 1) {
        std::cout << "CANNOT LOP HEAD" << std::endl;
        exit(2);
    }

    node* new_head = (*head)->next;
    if (new_head != nullptr) {
        delete *head;
        (*head) = new_head;
        if ((*head)->next != nullptr) {
            (*head)->next->heads += counter;
        }
    }
}

node* get_node_with_most_heads(node** head) {
    auto check_node = (*head);
    node *withMostHeads;
    int mostHeadsCount = 0;
    while (check_node != nullptr) {
        if (check_node->heads > mostHeadsCount) {
            mostHeadsCount = check_node->heads;
            withMostHeads = check_node;
        }
        check_node = check_node->next;
    }
    return withMostHeads;
}

void remove_head(node** n, unsigned long long int counter) {
    node* parent = (*n)->next;
    (*n)->heads--;
    if (parent != nullptr) {
        parent->heads += counter;
    }
}

bool has_heads(node node1) {
    return node1.next != nullptr;
}

int main() {
    unsigned long long int steps = 0;
    unsigned long long int previous_steps;
    int length = 5;

    node *head = nullptr;

    for (int i = 0; i <= length; i++) {
        insert_at_front(&head, 1);
    }

    while (has_heads(*head)) {
        if (are_all_heads_one(&head)) {
            steps++;
            lop_head(&head, steps);
        } else {
            steps++;
            node* withMostHeads = get_node_with_most_heads(&head);
            remove_head(&withMostHeads, steps);
        }
        if (steps < previous_steps) {
            std::cout << "overflow" << std::endl;
            exit(3);
        }
        previous_steps = steps;

    }

    std::cout << "steps: " << steps << std::endl;

    return 0;
}
